### DISCUSSION TITLE
Voice implementation

### DISCUSSION THREAD

**Discussion by Daniel-Boll** [uowuo/abaddon#80] [General]
Hello fellow discorders,

Do y'all already have a plan on how to handle the voice implementation?

**Comment by ouwou** [comment]
well ive figured out how to send voice and actually receive voice data, which is the scary undocumented stuff API-wise. but ive never done anything with audio in general though so perhaps ill just have to cross that bridge when i get there

**Reply by Daniel-Boll** [reply]
I'm looking forward to contribute with the project, probably mainly in the voice implementation, if anyone has any ideas on where to start researching would help.

Tried find a couple of related articles, yet still not successful.

The plan is to have the voice implementation independent of Discord?

**Reply by Daniel-Boll** [reply]
Because it seems to be possible that way https://discord.com/developers/docs/topics/voice-connections, otherwise create a bridge between the users in the voice chat themselves somehow, the problem would that this restrict voice chats on abaddon as virtual voice chats or something like that, which wouldn't reflect on actual Discord, and I think that this not the plan

**Reply by ouwou** [reply]
> Because it seems to be possible that way https://discord.com/developers/docs/topics/voice-connections, otherwise create a bridge between the users in the voice chat themselves somehow, the problem would that this restrict voice chats on abaddon as virtual voice chats or something like that, which wouldn't reflect on actual Discord, and I think that this not the plan

the documentation in that link is how you would do it. the undocumented part is receiving voice data, which you can do by sending the voice data `[0xF8, 0xFF, 0xFE]`. then you receive RTP packets and decrypt/decode the same way you encrypt/encode. you use the SSRC in the RTP header to figure out which user the audio is coming from since that information is sent over the voice websocket

if you want some examples of how others implemented it, most popular libraries have implemented sending voice data, and discord.js and one of the c# libraries (i forget which) has receiving voice too

**Reply by Daniel-Boll** [reply]
Okay, I'll play with that in the next few days, any progress I'll report here :lab_coat: :microscope:

**Reply by ouwou** [reply]
> Okay, I'll play with that in the next few days, any progress I'll report here 🥼 🔬

sure and if you run into any trouble or want to get into more details you can join the discord too
